File: medium/2685-count-the-number-of-complete-components/solution.java
import java.util.HashMap;
import java.util.Map;

/**
 * 2685. Count the Number of Complete Components
 * Time: O(n + m * alpha(n))
 * Space: O(n)
 */
class Solution {
    private int[] parent;
    private int[] size;

    public int countCompleteComponents(int n, int[][] edges) {
        parent = new int[n];
        size = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            size[i] = 1;
        }

        for (int[] e : edges) {
            union(e[0], e[1]);
        }

        // Count edges per component root.
        Map<Integer, Integer> edgeCount = new HashMap<>();
        for (int[] e : edges) {
            int r = find(e[0]);
            edgeCount.merge(r, 1, Integer::sum);
        }

        int result = 0;
        for (int v = 0; v < n; v++) {
            if (find(v) == v) { // v is a root
                long k = size[v];
                if (edgeCount.getOrDefault(v, 0) == k * (k - 1) / 2) {
                    result++;
                }
            }
        }
        return result;
    }

    private int find(int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }

    private void union(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra != rb) {
            if (size[ra] < size[rb]) {
                int t = ra;
                ra = rb;
                rb = t;
            }
            parent[rb] = ra;
            size[ra] += size[rb];
        }
    }
}

File: medium/2685-count-the-number-of-complete-components/solution.py
from typing import List


class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        """
        2685. Count the Number of Complete Components
        Time: O(n + m * alpha(n))
        Space: O(n)
        """
        parent = list(range(n))
        size = [1] * n

        def find(x: int) -> int:
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x

        def union(a: int, b: int) -> None:
            ra, rb = find(a), find(b)
            if ra != rb:
                if size[ra] < size[rb]:
                    ra, rb = rb, ra
                parent[rb] = ra
                size[ra] += size[rb]

        for a, b in edges:
            union(a, b)

        # Count edges per component root.
        edge_count = {}
        for a, b in edges:
            r = find(a)
            edge_count[r] = edge_count.get(r, 0) + 1

        result = 0
        for v in range(n):
            if find(v) == v:  # v is a root
                k = size[v]
                if edge_count.get(v, 0) == k * (k - 1) // 2:
                    result += 1
        return result
